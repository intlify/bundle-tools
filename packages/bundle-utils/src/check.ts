export type InstalledPackage = 'vue-i18n' | 'petite-vue-i18n'

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkInstallPackage(
  pkg: string,
  debug: Function
): InstalledPackage {
  let installedVueI18n = false
  try {
    debug(`vue-i18n load path: ${require.resolve('vue-i18n')}`)
    installedVueI18n = true
  } catch (e) {
    debug(`cannot find 'vue-i18n'`, e)
  }

  let installedPetiteVueI18n = false
  try {
    debug(`petite-vue-i18n load path: ${require.resolve('petite-vue-i18n')}`)
    installedPetiteVueI18n = true
  } catch (e) {
    debug(`cannot find 'petite-vue-i18n'`, e)
  }

  if (installedVueI18n) {
    return 'vue-i18n'
  }
  if (installedPetiteVueI18n) {
    return 'petite-vue-i18n'
  }
  throw new Error(
    `${pkg} requires 'vue-i18n' or 'petite-vue-i18n' to be present in the dependency tree.`
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkVueI18nBridgeInstallPackage(debug: Function): boolean {
  let ret = false
  try {
    debug(`vue-i18n-bridge load path: ${require.resolve('vue-i18n-bridge')}`)
    ret = true
  } catch (e) {
    debug(`cannot find 'vue-i18n-bridge'`, e)
  }
  return ret
}
