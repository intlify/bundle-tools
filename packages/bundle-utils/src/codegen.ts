import module from 'node:module'
import {
  SourceMapGenerator,
  SourceMapConsumer,
  MappedPosition,
  MappingItem
} from 'source-map'
import { format, escapeHtml as sanitizeHtml, isBoolean } from '@intlify/shared'

import type { RawSourceMap } from 'source-map'
import type {
  CompileError,
  ResourceNode,
  CompileOptions
} from '@intlify/message-compiler'

// NOTE: use sourcemap (currently, `@intlify/message-compiler` support cjs pkg only)
const require = module.createRequire(import.meta.url)
const {
  baseCompile,
  detectHtmlTag,
  LocationStub
} = require('@intlify/message-compiler')

/**
 * Compilation dev environments
 *
 * @public
 */
export type DevEnv = 'development' | 'production'

export interface Position {
  line: number
  column: number
  offset?: number
}

export interface SourceLocationable {
  start?: number
  loc?: {
    start: Position
    end: Position
  } | null
}

/**
 * @internal
 */
export interface CodeGenOptions {
  type?: 'plain' | 'sfc' | 'bare'
  legacy?: boolean
  bridge?: boolean
  exportESM?: boolean
  source?: string
  sourceMap?: boolean
  filename?: string
  inSourceMap?: RawSourceMap
  isGlobal?: boolean
  locale?: string
  env?: DevEnv
  forceStringify?: boolean
  useClassComponent?: boolean
  allowDynamic?: boolean
  strictMessage?: boolean
  escapeHtml?: boolean
  onWarn?: (msg: string) => void
  onError?: (
    msg: string,
    extra?: {
      source: string
      path: string
      code?: CompileError['code']
      domain?: CompileError['domain']
      location?: CompileError['location']
    }
  ) => void
}

export interface CodeGenContext {
  source?: string
  code: string
  indentLevel: number
  filename: string
  line: number
  column: number
  offset: number
  env: DevEnv
  forceStringify: boolean
  map?: SourceMapGenerator
}

export interface CodeGenerator {
  context(): CodeGenContext
  push<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void
  indent(withNewLine?: boolean): void
  deindent(withNewLine?: boolean): void
  newline(): void
  pushline<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void
}

/**
 * @internal
 */
export interface CodeGenResult<ASTNode, CodeGenError extends Error = Error> {
  code: string
  ast: ASTNode
  errors?: CodeGenError[]
  map?: RawSourceMap
}

export function createCodeGenerator(
  options: CodeGenOptions = {
    filename: 'bundle.json',
    sourceMap: false,
    env: 'development',
    forceStringify: false
  }
): CodeGenerator {
  const { sourceMap, source, filename } = options
  const _context = Object.assign(
    {
      code: '',
      column: 1,
      line: 1,
      offset: 0,
      map: undefined,
      indentLevel: 0
    },
    options
  ) as CodeGenContext

  const context = (): CodeGenContext => _context

  function push<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void {
    _context.code += code
    if (_context.map && node) {
      if (node.loc && node.loc !== LocationStub) {
        addMapping(node.loc.start, name)
      }
      advancePositionWithSource(_context as Position, code)
    }
  }

  function _newline(n: number): void {
    push('\n' + `  `.repeat(n))
  }

  function indent(withNewLine = true): void {
    const level = ++_context.indentLevel
    withNewLine && _newline(level)
  }

  function deindent(withNewLine = true): void {
    const level = --_context.indentLevel
    withNewLine && _newline(level)
  }

  function newline(): void {
    _newline(_context.indentLevel)
  }

  function pushline<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void {
    push(code, node, name)
    newline()
  }

  function addMapping(loc: Position, name?: string) {
    _context.map!.addMapping({
      name,
      source: _context.filename,
      original: {
        line: loc.line,
        column: loc.column - 1
      },
      generated: {
        line: _context.line,
        column: _context.column - 1
      }
    })
  }

  if (sourceMap && source) {
    _context.map = new SourceMapGenerator()
    _context.map.setSourceContent(filename!, source)
  }

  return {
    context,
    push,
    indent,
    deindent,
    newline,
    pushline
  }
}

function advancePositionWithSource(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  if (pos.offset == null) {
    return pos
  }

  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}

const DETECT_MESSAGE = `Detected HTML in '{msg}' message.`
const ON_ERROR_NOOP = () => {} // eslint-disable-line @typescript-eslint/no-empty-function

function parsePath(path?: string[]): string {
  return path ? path.join('.') : ''
}

export function generateMessageFunction(
  msg: string,
  options: CodeGenOptions,
  path?: string[]
): CodeGenResult<ResourceNode> {
  const env = options.env != null ? options.env : 'development'
  const strictMessage = isBoolean(options.strictMessage)
    ? options.strictMessage
    : true
  const escapeHtml = !!options.escapeHtml
  const onError = options.onError || ON_ERROR_NOOP
  const errors = [] as CompileError[]

  let detecteHtmlInMsg = false
  if (detectHtmlTag(msg)) {
    detecteHtmlInMsg = true
    if (strictMessage) {
      const errMsg = format(DETECT_MESSAGE, { msg })
      onError(format(errMsg), {
        source: msg,
        path: parsePath(path)
      })
    }
  }

  const _msg = detecteHtmlInMsg && escapeHtml ? sanitizeHtml(msg) : msg

  const newOptions = Object.assign({ mode: 'arrow' }, options) as CompileOptions
  newOptions.onError = (err: CompileError): void => {
    if (onError) {
      const extra: Parameters<Required<CodeGenOptions>['onError']>[1] = {
        source: msg,
        path: parsePath(path),
        code: err.code,
        domain: err.domain,
        location: err.location
      }
      onError(err.message, extra)
      errors.push(err)
    }
  }
  const { code, ast, map } = baseCompile(_msg, newOptions)
  const occured = errors.length > 0
  const genCode = !occured
    ? env === 'development'
      ? `(()=>{const fn=${code};fn.source=${JSON.stringify(msg)};return fn;})()`
      : `${code}`
    : _msg
  return { code: genCode, ast, map, errors }
}

export function mapLinesColumns(
  resMap: RawSourceMap,
  codeMaps: Map<string, RawSourceMap>,
  inSourceMap?: RawSourceMap
): RawSourceMap | null {
  if (!resMap) {
    return null
  }

  const resMapConsumer = new SourceMapConsumer(resMap)
  const inMapConsumer = inSourceMap ? new SourceMapConsumer(inSourceMap) : null
  const mergedMapGenerator = new SourceMapGenerator()

  let inMapFirstItem: MappingItem | null = null
  if (inMapConsumer) {
    inMapConsumer.eachMapping(m => {
      if (inMapFirstItem) {
        return
      }
      inMapFirstItem = m
    })
  }

  resMapConsumer.eachMapping(res => {
    if (res.originalLine == null) {
      return
    }

    const map = codeMaps.get(res.name)
    if (!map) {
      return
    }

    let inMapOrigin: MappedPosition | null = null
    if (inMapConsumer) {
      inMapOrigin = inMapConsumer.originalPositionFor({
        line: res.originalLine,
        column: res.originalColumn - 1
      })
      if (inMapOrigin.source == null) {
        inMapOrigin = null
        return
      }
    }

    const mapConsumer = new SourceMapConsumer(map)
    mapConsumer.eachMapping(m => {
      mergedMapGenerator.addMapping({
        original: {
          line: inMapFirstItem
            ? inMapFirstItem.originalLine + res.originalLine - 2
            : res.originalLine,
          column: inMapFirstItem
            ? inMapFirstItem.originalColumn + res.originalColumn
            : res.originalColumn
        },
        generated: {
          line: inMapFirstItem
            ? inMapFirstItem.generatedLine + res.originalLine - 2
            : res.originalLine,
          // map column with message format compilation code map
          column: inMapFirstItem
            ? inMapFirstItem.generatedColumn +
              res.originalColumn +
              m.generatedColumn
            : res.originalColumn + m.generatedColumn
        },
        source: inMapOrigin ? inMapOrigin.source : res.source,
        name: m.name // message format compilation code
      })
    })
  })

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const generator = mergedMapGenerator as any
  // const targetConsumer = inMapConsumer || resMapConsumer
  const targetConsumer = inMapConsumer || resMapConsumer
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(targetConsumer as any).sources.forEach((sourceFile: string) => {
    generator._sources.add(sourceFile)
    const sourceContent = targetConsumer.sourceContentFor(sourceFile)
    if (sourceContent != null) {
      mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
    }
  })

  generator._sourceRoot = inSourceMap
    ? inSourceMap.sourceRoot
    : resMap.sourceRoot
  generator._file = inSourceMap ? inSourceMap.file : resMap.file

  return generator.toJSON()
}
