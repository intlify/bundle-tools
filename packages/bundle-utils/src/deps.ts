import module from 'node:module'

export type InstalledPackage = 'vue-i18n' | 'petite-vue-i18n'

const _require = module.createRequire(import.meta.url)

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkInstallPackage(
  pkg: string,
  debug: Function
): InstalledPackage {
  let installedVueI18n = false
  try {
    debug(`vue-i18n load path: ${_require.resolve('vue-i18n')}`)
    installedVueI18n = true
  } catch (e) {
    debug(`cannot find 'vue-i18n'`, e)
  }

  let installedPetiteVueI18n = false
  try {
    debug(`petite-vue-i18n load path: ${_require.resolve('petite-vue-i18n')}`)
    installedPetiteVueI18n = true
  } catch (e) {
    debug(`cannot find 'petite-vue-i18n'`, e)
  }

  if (installedVueI18n) {
    return 'vue-i18n'
  }
  if (installedPetiteVueI18n) {
    return 'petite-vue-i18n'
  }
  throw new Error(
    `${pkg} requires 'vue-i18n' or 'petite-vue-i18n' to be present in the dependency tree.`
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkVueI18nBridgeInstallPackage(debug: Function): boolean {
  let ret = false
  try {
    debug(`vue-i18n-bridge load path: ${_require.resolve('vue-i18n-bridge')}`)
    ret = true
  } catch (e) {
    debug(`cannot find 'vue-i18n-bridge'`, e)
  }
  return ret
}

type VueI18nVersion = '8' | '9' | 'unknown' | ''

export function getVueI18nVersion(debug: Function): VueI18nVersion {
  const VueI18n = loadModule('vue-i18n', debug)
  if (VueI18n == null) {
    return ''
  }
  if (VueI18n.version && VueI18n.version.startsWith('8.')) {
    return '8'
  }
  if (VueI18n.VERSION && VueI18n.VERSION.startsWith('9.')) {
    return '9'
  }
  return 'unknown'
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function loadModule(moduleName: string, debug: Function): any {
  try {
    return _require(moduleName)
  } catch (e) {
    debug(`cannot load '${moduleName}'`, e)
    return null
  }
}
