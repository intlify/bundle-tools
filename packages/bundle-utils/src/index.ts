export { generate as generateJSON } from './json'
export { generate as generateYAML } from './yaml'
export { generate as generateJavaScript } from './js'
export {
  getVueI18nVersion,
  checkInstallPackage,
  checkVueI18nBridgeInstallPackage
} from './deps'
export type { CodeGenOptions, CodeGenResult, DevEnv } from './codegen'
export type { InstalledPackage } from './deps'
