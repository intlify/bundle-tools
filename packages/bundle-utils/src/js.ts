/**
 * Code generator for i18n js resource
 */

import { isString, isBoolean, isNumber } from '@intlify/shared'
import { parse as parseJavaScript } from 'acorn'
import { generate as generateJavaScript } from 'escodegen'
import { walk } from 'estree-walker'
import {
  createCodeGenerator,
  generateMessageFunction,
  mapLinesColumns
} from './codegen'
import { RawSourceMap } from 'source-map'

import type { Node } from 'estree'
import type { CodeGenOptions, CodeGenerator, CodeGenResult } from './codegen'

/**
 * @internal
 */
export function generate(
  targetSource: string | Buffer,
  {
    type = 'plain',
    bridge = false,
    exportESM = false,
    filename = 'vue-i18n-loader.js',
    inSourceMap = undefined,
    locale = '',
    isGlobal = false,
    sourceMap = false,
    env = 'development',
    forceStringify = false,
    onError = undefined,
    strictMessage = true,
    escapeHtml = false,
    useClassComponent = false,
    allowDynamic = false
  }: CodeGenOptions,
  injector?: () => string
): CodeGenResult<Node> {
  const target = Buffer.isBuffer(targetSource)
    ? targetSource.toString()
    : targetSource
  const value = target

  const options = {
    type,
    bridge,
    exportESM,
    source: value,
    sourceMap,
    locale,
    isGlobal,
    inSourceMap,
    env,
    filename,
    forceStringify,
    onError,
    strictMessage,
    escapeHtml,
    useClassComponent
  } as CodeGenOptions
  const generator = createCodeGenerator(options)

  const ast = parseJavaScript(value, {
    ecmaVersion: 'latest',
    sourceType: 'module',
    sourceFile: filename,
    allowImportExportEverywhere: true
  }) as Node

  const exportResult = scanAst(ast)
  if (!allowDynamic) {
    // if (!astExportDefaultWithObject.length) {
    if (!exportResult || exportResult !== 'object') {
      throw new Error(
        `You need to define an object as the locale message with 'export default'.`
      )
    }
  } else {
    if (!exportResult) {
      throw new Error(
        `You need to define 'export default' that will return the locale messages.`
      )
    }

    if (exportResult !== 'object') {
      /**
       * NOTE:
       *  If `allowDynamic` is `true`, do not transform the code by this function, return it as is.
       *  This means that the user **must transform locale messages ownself**.
       *  Especially at the production, you need to do locale messages pre-compiling.
       */
      return {
        ast,
        code: value,
        map: inSourceMap
      }
    }
  }

  const codeMaps = generateNode(generator, ast, options, injector)

  const { code, map } = generator.context()
  // if (map) {
  //   const s = new SourceMapConsumer((map as any).toJSON())
  //   s.eachMapping(m => {
  //     console.log('sourcemap json', m)
  //   })
  // }
  // prettier-ignore
  const newMap = map
    ? mapLinesColumns((map as any).toJSON(), codeMaps, inSourceMap) || null // eslint-disable-line @typescript-eslint/no-explicit-any
    : null
  return {
    ast,
    code,
    map: newMap != null ? newMap : undefined
  }
}

function scanAst(ast: Node) {
  if (ast.type !== 'Program') {
    throw new Error('Invalid AST: does not have Program node')
  }

  let ret: false | 'object' | 'function' | 'arrow-function' = false
  for (const node of ast.body) {
    if (node.type === 'ExportDefaultDeclaration') {
      if (node.declaration.type === 'ObjectExpression') {
        ret = 'object'
        break
      } else if (node.declaration.type === 'FunctionDeclaration') {
        ret = 'function'
        break
      } else if (node.declaration.type === 'ArrowFunctionExpression') {
        ret = 'arrow-function'
        break
      }
    }
  }
  return ret
}

function generateNode(
  generator: CodeGenerator,
  node: Node,
  options: CodeGenOptions,
  injector?: () => string
): Map<string, RawSourceMap> {
  const propsCountStack = [] as number[]
  const pathStack = [] as string[]
  const itemsCountStack = [] as number[]
  const skipStack = [] as boolean[]
  const { forceStringify } = generator.context()
  const codeMaps = new Map<string, RawSourceMap>()
  const {
    type,
    bridge,
    exportESM,
    sourceMap,
    isGlobal,
    locale,
    useClassComponent
  } = options

  const componentNamespace = '_Component'

  walk(node, {
    /**
     * NOTE:
     *  force cast to Node of `estree-walker@3.x`,
     *  because `estree-walker@3.x` is not dual packages,
     *  so it's support only esm only ...
     */
    // @ts-ignore
    enter(node: Node, parent: Node) {
      switch (node.type) {
        case 'Program':
          if (type === 'plain') {
            generator.push(`const resource = `)
          } else if (type === 'sfc') {
            // for 'sfc'
            const variableName =
              type === 'sfc' ? (!isGlobal ? '__i18n' : '__i18nGlobal') : ''
            const localeName =
              type === 'sfc' ? (locale != null ? locale : `""`) : ''
            const exportSyntax = bridge
              ? exportESM
                ? `export default`
                : `module.exports =`
              : `export default`
            generator.push(`${exportSyntax} function (Component) {`)
            generator.indent()
            // prettier-ignore
            const componentVariable = bridge
              ? `Component.options || Component`
              : useClassComponent
                ? `Component.__o || Component.__vfdConstructor || Component`
                : `Component`
            // prettier-ignore
            generator.pushline(`const ${componentNamespace} = ${componentVariable}`)
            generator.pushline(
              `${componentNamespace}.${variableName} = ${componentNamespace}.${variableName} || []`
            )
            generator.push(`${componentNamespace}.${variableName}.push({`)
            generator.indent()
            generator.pushline(`"locale": ${JSON.stringify(localeName)},`)
            generator.push(`"resource": `)
          }
          break
        case 'ObjectExpression':
          generator.push(`{`)
          generator.indent()
          propsCountStack.push(node.properties.length)
          if (parent != null && parent.type === 'ArrayExpression') {
            const lastIndex = itemsCountStack.length - 1
            const currentCount =
              parent.elements.length - itemsCountStack[lastIndex]
            pathStack.push(currentCount.toString())
            itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
          }
          break
        case 'Property':
          if (parent != null && parent.type === 'ObjectExpression') {
            if (node != null) {
              if (
                isJSONablePrimitiveLiteral(node.value) &&
                (node.key.type === 'Literal' || node.key.type === 'Identifier')
              ) {
                // prettier-ignore
                const name = node.key.type === 'Literal'
                ? String(node.key.value)
                : node.key.name
                if (
                  (node.value.type === 'Literal' &&
                    isString(node.value.value)) ||
                  node.value.type === 'TemplateLiteral'
                ) {
                  const value = getValue(node.value) as string
                  generator.push(`${JSON.stringify(name)}: `)
                  pathStack.push(name)
                  const { code, map } = generateMessageFunction(
                    value,
                    options,
                    pathStack
                  )
                  sourceMap && map != null && codeMaps.set(value, map)
                  generator.push(`${code}`, node.value, value)
                  skipStack.push(false)
                } else {
                  const value = getValue(node.value)
                  if (forceStringify) {
                    const strValue = JSON.stringify(value)
                    generator.push(`${JSON.stringify(name)}: `)
                    pathStack.push(name)
                    const { code, map } = generateMessageFunction(
                      strValue,
                      options,
                      pathStack
                    )
                    sourceMap && map != null && codeMaps.set(strValue, map)
                    generator.push(`${code}`, node.value, strValue)
                  } else {
                    generator.push(
                      `${JSON.stringify(name)}: ${JSON.stringify(value)}`
                    )
                    pathStack.push(name)
                  }
                  skipStack.push(false)
                }
              } else if (
                (node.value.type === 'FunctionExpression' ||
                  node.value.type === 'ArrowFunctionExpression') &&
                (node.key.type === 'Literal' || node.key.type === 'Identifier')
              ) {
                // prettier-ignore
                const name = node.key.type === 'Literal'
                ? String(node.key.value)
                : node.key.name
                generator.push(`${JSON.stringify(name)}: `)
                pathStack.push(name)
                const code = generateJavaScript(node.value)
                generator.push(`${code}`, node.value, code)
                skipStack.push(false)
              } else if (
                (node.value.type === 'ObjectExpression' ||
                  node.value.type === 'ArrayExpression') &&
                (node.key.type === 'Literal' || node.key.type === 'Identifier')
              ) {
                // prettier-ignore
                const name = node.key.type === 'Literal'
                ? String(node.key.value)
                : node.key.name
                generator.push(`${JSON.stringify(name)}: `)
                pathStack.push(name)
              } else {
                // for Regex, function, etc.
                skipStack.push(true)
              }
            }
            const lastIndex = propsCountStack.length - 1
            propsCountStack[lastIndex] = --propsCountStack[lastIndex]
          }
          break
        case 'ArrayExpression':
          generator.push(`[`)
          generator.indent()
          if (parent != null && parent.type === 'ArrayExpression') {
            const lastIndex = itemsCountStack.length - 1
            const currentCount =
              parent.elements.length - itemsCountStack[lastIndex]
            pathStack.push(currentCount.toString())
            itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
          }
          itemsCountStack.push(node.elements.length)
          break
        default:
          if (node != null && parent != null) {
            if (parent.type === 'ArrayExpression') {
              const lastIndex = itemsCountStack.length - 1
              const currentCount =
                parent.elements.length - itemsCountStack[lastIndex]
              pathStack.push(currentCount.toString())
              if (isJSONablePrimitiveLiteral(node)) {
                if (
                  (node.type === 'Literal' && isString(node.value)) ||
                  node.type === 'TemplateLiteral'
                ) {
                  const value = getValue(node) as string
                  const { code, map } = generateMessageFunction(
                    value,
                    options,
                    pathStack
                  )
                  sourceMap && map != null && codeMaps.set(value, map)
                  generator.push(`${code}`, node, value)
                } else {
                  const value = getValue(node)
                  if (forceStringify) {
                    const strValue = JSON.stringify(value)
                    const { code, map } = generateMessageFunction(
                      strValue,
                      options,
                      pathStack
                    )
                    sourceMap && map != null && codeMaps.set(strValue, map)
                    generator.push(`${code}`, node, strValue)
                  } else {
                    generator.push(`${JSON.stringify(value)}`)
                  }
                }
                skipStack.push(false)
              } else {
                // for Regex, function, etc.
                skipStack.push(true)
              }
              itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
            }
          } else {
            // ...
          }
          break
      }
    },
    /**
     * NOTE:
     *  force cast to Node of `estree-walker@3.x`,
     *  because `estree-walker@3.x` is not dual packages,
     *  so it's support only esm only ...
     */
    // @ts-ignore
    leave(node: Node, parent: Node) {
      switch (node.type) {
        case 'Program':
          if (type === 'sfc') {
            generator.deindent()
            generator.push(`})`)
            if (bridge && injector) {
              generator.newline()
              generator.pushline(
                `${componentNamespace}.__i18nBridge = ${componentNamespace}.__i18nBridge || []`
              )
              generator.pushline(
                `${componentNamespace}.__i18nBridge.push('${injector()}')`
              )
              generator.pushline(`delete ${componentNamespace}._Ctor`)
            }
            generator.deindent()
            generator.pushline(`}`)
          } else if (type === 'plain') {
            generator.push(`\n`)
            generator.push('export default resource')
          }
          break
        case 'ObjectExpression':
          if (propsCountStack[propsCountStack.length - 1] === 0) {
            pathStack.pop()
            propsCountStack.pop()
          }
          generator.deindent()
          generator.push(`}`)
          if (parent != null && parent.type === 'ArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              generator.pushline(`,`)
            }
          }
          break
        case 'Property':
          if (parent != null && parent.type === 'ObjectExpression') {
            if (propsCountStack[propsCountStack.length - 1] !== 0) {
              pathStack.pop()
              if (!skipStack.pop()) {
                generator.pushline(`,`)
              }
            }
          }
          break
        case 'ArrayExpression':
          if (itemsCountStack[itemsCountStack.length - 1] === 0) {
            pathStack.pop()
            itemsCountStack.pop()
          }
          generator.deindent()
          generator.push(`]`)
          if (parent != null && parent.type === 'ArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              if (!skipStack.pop()) {
                generator.pushline(`,`)
              }
            }
          }
          break
        case 'Literal':
          if (parent != null && parent.type === 'ArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              if (!skipStack.pop()) {
                generator.pushline(`,`)
              }
            } else {
              if (!skipStack.pop()) {
                generator.pushline(`,`)
              }
            }
          }
          break
        default:
          break
      }
    }
  })

  return codeMaps
}

function isJSONablePrimitiveLiteral(node: Node): boolean {
  return (
    (node.type === 'Literal' &&
      (isString(node.value) ||
        isNumber(node.value) ||
        isBoolean(node.value) ||
        node.value === null)) ||
    node.type === 'TemplateLiteral'
  )
  // NOTE: the following code is same the above code
  /*
  if (node.type === 'Literal') {
    if (
      isString(node.value) ||
      isNumber(node.value) ||
      isBoolean(node.value) ||
      node.value === null
    ) {
      return true
    } else if (isRegExp(node.value)) {
      return false
    } else {
      return false
    }
  } else if (node.type === 'TemplateLiteral') {
    return true
  } else {
    return false
  }
  */
}

function getValue(node: Node) {
  // prettier-ignore
  return node.type === 'Literal'
    ? node.value
    : node.type === 'TemplateLiteral'
      ? node.quasis.map(quasi => quasi.value.cooked).join('')
      : undefined
}
