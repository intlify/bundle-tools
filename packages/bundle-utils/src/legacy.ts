import { friendlyJSONstringify } from '@intlify/shared'

import type { CodeGenOptions } from './codegen'

export function generateLegacyCode(
  { isGlobal }: CodeGenOptions,
  generator: () => string
) {
  // prettier-ignore
  const componentNamespace = `Component.options`
  const variableName = !isGlobal ? '__i18n' : '__i18nGlobal'
  const exportSyntax = 'export default'
  const code = `${exportSyntax} function (Component) {
  ${componentNamespace}.${variableName} = ${componentNamespace}.${variableName} || []
  ${componentNamespace}.${variableName}.push(${friendlyJSONstringify(
    generator()
  )})
  delete ${componentNamespace}._Ctor
}`
  return code
}
