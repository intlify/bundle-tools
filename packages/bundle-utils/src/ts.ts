/**
 * Code generator for i18n ts resource
 *
 * An async wrapper for the js generator which transforms ts to js before generating
 */

import { transform } from 'esbuild'
import { Node } from 'estree'
import { CodeGenOptions, CodeGenResult } from './codegen'
import { DEFAULT_OPTIONS, generate as generateJavaScript } from './js'

/**
 * @internal
 */
export async function generate(
  targetSource: string | Buffer,
  options: CodeGenOptions
): Promise<CodeGenResult<Node>> {
  let value = Buffer.isBuffer(targetSource) ? targetSource.toString() : targetSource

  const _options = Object.assign({}, DEFAULT_OPTIONS, options, { source: value })
  if (_options.filename && /.[c|m]?ts$/.test(_options.filename)) {
    const transformed = await transform(value, { loader: 'ts' })

    if (transformed && transformed.code) {
      value = transformed.code
      _options.source = transformed.code
    }
  }

  return generateJavaScript(value, _options)
}
