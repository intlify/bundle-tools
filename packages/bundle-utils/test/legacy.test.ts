import { generateLegacyCode } from '../src/legacy'

test('not global', () => {
  expect(
    generateLegacyCode({ isGlobal: false }, () =>
      JSON.stringify({ foo: 'bar' })
    )
  ).toMatch('Component.options.__i18n =')
})

test('global', () => {
  expect(
    generateLegacyCode({ isGlobal: true }, () => JSON.stringify({ foo: 'bar' }))
  ).toMatch('Component.options.__i18nGlobal =')
})

test('vue 2.7', () => {
  expect(
    generateLegacyCode({ vueVersion: 'v2.7' }, () =>
      JSON.stringify({ foo: 'bar' })
    )
  ).toMatch('Component.__i18n =')
})
