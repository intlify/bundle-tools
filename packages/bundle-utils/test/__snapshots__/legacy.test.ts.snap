// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`global 1`] = `
"export default function (Component) {
  Component.options.__i18nGlobal = Component.options.__i18nGlobal || []
  Component.options.__i18nGlobal.push(\\"{\\\\\\"foo\\\\\\":\\\\\\"bar\\\\\\"}\\")
  delete Component.options._Ctor
}"
`;
